Re-introduce terse configuration

akshayjshah opened the pull request:
This PR re-introduces a nice, terse way to configure sane-by-default loggers. It also introduces a configuration struct for loggers, since many applications want to drive logger configuration from config instead of code.

akshayjshah commented on zapcore/encoder.go:
```diff
@@ -80,16 +182,32 @@ type ObjectEncoder interface {
 // aren't safe for concurrent use (though typical use shouldn't require locks).
 type ArrayEncoder interface {
+	// Built-in types
+	PrimitiveArrayEncoder
```
I'll break out this change into a separate PR - it was an oversight on my part. Passing an `ArrayEncoder` to a `TimeEncoder` opens us up to infinite recursion :/

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+)
+
+// SamplingConfig sets a sampling strategy for the logger. Sampling caps the
+// global CPU and I/O load that logging puts on your process while preserving a
```
I'd prefer to soften this to "while attempting to preserve a".

akshayjshah replied:
:+1:

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	// Development puts the logger in development mode, which changes the
+	// behavior of DPanicLevel and takes stacktraces more liberally.
+	Development bool `json:"development",yaml:"development"`
+	// DisableCaller stops annotating logs with the calling function's file
```
Say more about when it happens if not enabled, always? Only for warn and up?

akshayjshah replied:
Good point.

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	// DisableCaller stops annotating logs with the calling function's file
+	// name and line number.
+	DisableCaller bool `json:"disable_caller",yaml:"disable_caller"`
+	// DisableStacktrace completely disables automatic stacktrace capturing.
```
Say more about when it happens if not enabled, always? Only for error and up?

prashantv replied:
Can you add information about when stack traces are captured (looks like error or higher for prod, and warn or higher for dev)

akshayjshah replied:
Great minds do indeed think alike. 👍  👍

akshayjshah replied:
:+1:

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	// level, so calling Config.Level.SetLevel will atomically change the log
+	// level of all loggers descended from this config. The zero value is
+	// InfoLevel.
+	Level AtomicLevel `json:"level",yaml:"level"`
```
How about a:
```
Level Level
AllowLevelChange bool
```

And defaulting to static levels?

akshayjshah replied:
Why? The performance overhead is near-zero, and I'd always recommend using a dynamic level.

prashantv commented on config.go:
```diff
@@ -0,0 +1,213 @@
+// global CPU and I/O load that logging puts on your process while preserving a
+// representative subset of your logs.
+//
+// See zapcore.Sample for details.
```
Should we mention that these values are per-second?

akshayjshah replied:
:+1:

prashantv commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	Development bool `json:"development",yaml:"development"`
+	// DisableCaller stops annotating logs with the calling function's file
+	// name and line number.
+	DisableCaller bool `json:"disable_caller",yaml:"disable_caller"`
```
Any reason for snake case in JSON? I tend to use the google JSON style guide which suggests `camelCase`
https://google.github.io/styleguide/jsoncstyleguide.xml#Property_Name_Format

I'm not sure what's appropriate for YAML, don't think there's any standard, I've seen `hyphernated-case` as well as `camelCase` and `snake_case`

akshayjshah replied:
Seems like everyone just ports the convention from their primary programming language. FB's API is snake-cased, Twitter's API is snake-cased, and the portions of the AWS APIs I've used are snake-cased.

I don't really care much though. Happy to change if you have a strong preference.

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	DisableStacktrace bool `json:"disable_stacktrace",yaml:"disable_stacktrace"`
+	// Sampling sets a sampling policy. A nil SamplingConfig disables sampling.
+	Sampling *SamplingConfig `json:"sampling",yaml:"sampling"`
+	// Encoding sets the logger's encoding. Valid values are "json" and
```
Interesting, I would've expected more of a:
```
EnableConsole bool // maybe disable instead of enable
DisableStandardErrors bool
FileEncoding string
OutputPaths []string
ErrorOutputPaths []string
```

The idea being:
- stdout is always console encoded
- files are always json encoded
- errors at least go to stderr, and maybe also a file

akshayjshah replied:
I'm glad that we can support that setup, but I'm less of a fan of making that part of the quickstart. I remembered discussing it with you and played with it for a while, and I found it pretty jarring. IMO it's just too unusual.

If and when we add new encoders, I'd also like this config struct to support them. Again, I have msgpack in production in mind.

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+		Encoding:    "console",
+		EncoderConfig: zapcore.EncoderConfig{
+			// Keys can be anything except the empty string.
+			TimeKey:        "T",
```
These are some surprising defaults imo, but I guess not that the console encoder uses them much?

Apropos of my above remarks, I think I'd still expect a "development" or "staging" deploy to log json to files, and I'd still want to see warn and up on stdout for things like docker logs in production...

akshayjshah replied:
The console encoder only cares about empty/non-empty - the actual values aren't used.

Terminology varies, but I'd call anything that's "deployed" a production config. In production, you may have canary, staging, and stable, but it's all production. This is for use while developers are writing code, running unit tests, and exploring on their local machines.

jcorbin replied:
Ahh okay, thanks for clarifying what "development" means.

akshayjshah replied:
Hm. Can't tell if that's mild sarcasm or not :)

Is there a better name we can use here? Not too late to rename "development" throughout the codebase.

jcorbin replied:
Nope, wasn't meant to be sarcastic at all, what I meant:
- to me, the word "development" in zap is 90% about "will DPanic panic?'
- I can see a sensible argument for wanting a staging or canary cluster to go either way on that point
- however I'd always expect a such-deployed cluster to always log to similar channels and encoded the same way as production, the only exception being when running tests, and otherwise "local"/non-deployed development

So I'm not sure if "development" is the best name or not for "non deployed config default", and whether or not we want the ability to vary DPanic's disposition within a canned "deployed config".

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+		return nil, err
+	}
+
+	enc, err := cfg.buildEncoder()
```
How about building the encoder first, before opening the sinks to minimize open file handles under error branch?

akshayjshah replied:
Sure.

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	return New(
+		zapcore.WriterFacility(enc, sink, cfg.Level),
+		cfg.buildOptions(errSink)...,
+	).WithOptions(opts...), nil
```
Much utility in not doing the `WithOptions` if empty?

akshayjshah replied:
Sure, can do - might as well save an encoder clone.

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+
+	if len(cfg.InitialFields) > 0 {
+		fs := make([]zapcore.Field, 0, len(cfg.InitialFields))
+		for k, v := range cfg.InitialFields {
```
Is that a `map[string]interface{}`? Shouldn't we provide a stable field order by sorting the keys?

akshayjshah replied:
Sure, can do.

jcorbin commented on config.go:
```diff
@@ -0,0 +1,213 @@
+		return nil, nil, err
+	}
+	return sink, errSink, nil
+}
```
I'd feel mildly better if we could ensure closing happens under errors... but actually doesn't Go setup a finalizer for that by default? Maybe my concerns are baseless?

akshayjshah replied:
Certainly possible; I'll just have the `Open` function also return a `close` closure. I suppose some people will want to continue operating even if they can't open their log files, so we should be careful.

jcorbin commented on config_test.go:
```diff
@@ -0,0 +1,86 @@
+			require.NoError(t, err, "Failed to create temp file.")
+			defer os.Remove(temp.Name())
+
+			tt.cfg.OutputPaths = []string{temp.Name()}
```
I always feel wary when I see a test fixture mutation...

akshayjshah replied:
Happy to assert that the field isn't set beforehand.

prashantv commented on config.go:
```diff
@@ -0,0 +1,213 @@
+			EncodeTime:     zapcore.EpochTimeEncoder,
+			EncodeDuration: zapcore.SecondsDurationEncoder,
+		},
+		OutputPaths:      []string{"stdout"},
```
Should the default output for logs be stderr?
logrus seems to use stderr by default: https://godoc.org/github.com/sirupsen/logrus#Logger
As does the standard go `log` package.

(Ideally production configs would actually use a file, but that's not a great default)

akshayjshah replied:
Sure, sounds good. This is yet another place where's Go chose to go down an unusual route :(

We'll have to disagree about prod logs - I still think they should log to standard out, and it's the process manager's job to pipe that to a file. That's the [12-factor standard](https://12factor.net/logs).

jcorbin commented on config_test.go:
```diff
@@ -0,0 +1,86 @@
+
+			tt.cfg.OutputPaths = []string{temp.Name()}
+			tt.cfg.EncoderConfig.TimeKey = "" // no timestamps in tests
+			tt.cfg.InitialFields = map[string]interface{}{"k": "v"}
```
And if you add a second field, this becomes unstable (due to map ordering per above comment)

akshayjshah replied:
Yep, good point.

prashantv commented on config.go:
```diff
@@ -0,0 +1,213 @@
+			EncodeTime:     zapcore.ISO8601TimeEncoder,
+			EncodeDuration: zapcore.StringDurationEncoder,
+		},
+		OutputPaths:      []string{"stdout"},
```
same as earlier, might want to use stderr for all logs

akshayjshah replied:
Yep.

jcorbin commented on level.go:
```diff
@@ -87,3 +87,19 @@ func (lvl AtomicLevel) Level() zapcore.Level {
+// "panic", and "fatal").
+func (lvl *AtomicLevel) UnmarshalText(text []byte) error {
+	if lvl.l == nil {
+		lvl.l = atomic.NewInt32(int32(InfoLevel))
```
Prefer to defer thus until after we've parsed without err.

jcorbin commented on level.go:
```diff
@@ -87,3 +87,19 @@ func (lvl AtomicLevel) Level() zapcore.Level {
+	if err := l.UnmarshalText(text); err != nil {
+		return err
+	}
+	lvl.SetLevel(l)
```
Prefer to:
```
if lvl.l == nil {
  lvl.l = atomic.NewInt32(int32(l))
} else {
  lvl.SetLevel(l)
}
```

akshayjshah replied:
:+1:

prashantv commented on config.go:
```diff
@@ -0,0 +1,213 @@
+
+	stackLevel := ErrorLevel
+	if cfg.Development {
+		stackLevel = WarnLevel
```
optional: can move the `Development()` option append here

akshayjshah replied:
Since we're handling a lot of different things in here, I'd rather handle each of the config struct's fields in a separate little block.

prashantv commented on config.go:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	if cfg.Sampling != nil {
+		opts = append(opts, WrapFacility(func(fac zapcore.Facility) zapcore.Facility {
```
(just a question) is there a difference between wrapping the facility when passing it into `New` vs calling `WrapFacility`?

akshayjshah replied:
Nope. This was just a bit easier organizationally - otherwise creating the facility needed its own helper method.

jcorbin commented on logger.go:
```diff
@@ -105,6 +99,16 @@ func (log *Logger) Named(s string) *Logger {
 }
 
+// WithOptions clones the current Logger, applies the supplied Options, and
+// returns the result. It's safe to use concurrently.
```
"... and returns the clone." (... "cloned child logger." ?)

akshayjshah replied:
Done.

prashantv commented on writer.go:
```diff
@@ -0,0 +1,64 @@
+		return zapcore.AddSync(ioutil.Discard), nil
+	}
+	writers, err := open(paths)
+	return zapcore.Lock(zapcore.MultiWriteSyncer(writers...)), err
```
Should `MultiWriteSyncer` always lock?

akshayjshah replied:
Not sure exactly what you're asking here. Since we're opening paths, the WriteSyncers in question are files, so we always need a lock.

In general, though, I don't think MultiWriteSyncers should lock.

prashantv commented on writer.go:
```diff
@@ -0,0 +1,64 @@
+func Open(paths ...string) (zapcore.WriteSyncer, error) {
+	if len(paths) == 0 {
+		return zapcore.AddSync(ioutil.Discard), nil
+	}
```
can we special case `len(paths) == 1` and return the `WriteSyncer` directly

akshayjshah replied:
:+1:

jcorbin commented on writer.go:
```diff
@@ -0,0 +1,64 @@
+			writers = append(writers, os.Stderr)
+			continue
+		}
+		f, err := os.OpenFile(path, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0644)
```
if this were a `default:`, you wouldn't need those `continue`s

prashantv replied:
Since we never read the file, I think we can use `O_WRONLY`.

For the mask, it is then umasked, so create uses `0666`:
https://golang.org/src/os/file.go?s=7623:7662#L248

We should probably do the same

akshayjshah replied:
Yes, but then I'd need another level of nesting. To me, this is similar to an early return. Reasonable?

akshayjshah replied:
:+1:

prashantv commented on zapcore/sampler.go:
```diff
@@ -65,7 +65,13 @@ func fnv32a(s string) uint32 {
-// Sample creates a facility that samples incoming entries.
+// Sample creates a Facility that samples incoming entries, which caps the CPU
+// and I/O load of logging while preserving a representative subset of your
+// logs.
```
should we mention the "optimized for performance over correctness" here as well?

akshayjshah replied:
Sure. I was saving that until the sampler fix lands, but this is a perfectly fine time to add it.

jcorbin commented on zapcore/sampler.go:
```diff
@@ -65,7 +65,13 @@ func fnv32a(s string) uint32 {
 
-// Sample creates a facility that samples incoming entries.
+// Sample creates a Facility that samples incoming entries, which caps the CPU
+// and I/O load of logging while preserving a representative subset of your
```
nit: "attempts to preserve a"

akshayjshah replied:
:+1:

akshayjshah commented:
Addressed all comments except:

1. Casing of JSON/YAML keys in config struct. Not opposed to changing, but the widely-used APIs I'm familiar with prefer snake case.
2. Different encodings for different outputs. In practice, feels jarring to me; also complicates config struct and future extensibility story.
